### PR TITLE
stop using --name option

### PR DESCRIPTION
**Pull request by akm**
use Docker HTTP API instead of docker command to start and stop container.

It's possible to run same typed containers at the same time.
## reviewers
- [x] @nagachika 
- [x] @minimum2scp

### CODE REVIEW

**Comment by nagachika** [comment]
Can this run with docker daemon listening the Unix Domain Socket? In other words, doesn't this pull request drop the support for docker with default setting (without DOCKER_HOST)?

Though I'm okey to accept the limitation myself, because I'm using boot2docker on my laptop, it should be noted on the README.md if this introduce such a precondition.

File: lib/mcrain/riak.rb [hunk 1]
```diff
@@ -1,27 +1,84 @@
...
+
+      def http_port
+        @http_port ||= find_portno
+      end
 
-    self.container_image = nil # not use docker directly
-    self.port = 8087
+      def build_docker_options
+        r = super
+        r['HostConfig']['PortBindings']["8098/tcp"] = [{ 'HostPort': http_port.to_s }]
```
Comment (minimum2scp) [inline lib/mcrain/riak.rb]: `[{ 'HostPort': http_port.to_s }]` should be `[{ 'HostPort' => http_port.to_s }]` (for Ruby 2.1.x)

**Comment by minimum2scp** [comment]
Example "Mcrain::Riak allow duplicated" fails.

```
yamada@MIRA% export DOCKER_HOST=tcp://172.17.42.1:2375
yamada@MIRA% bundle exec rake spec
/usr/bin/ruby2.1 -I/home/yamada/go/src/github.com/groovenauts/mcrain/vendor/bundle/ruby/2.1.0/gems/rspec-core-3.3.1/lib:/home/yamada/go/src/github.com/groovenauts/mcrain/vendor/bundle/ruby/2.1.0/gems/rspec-support-3.3.0/lib /home/yamada/go/src/github.com/groovenauts/mcrain/vendor/bundle/ruby/2.1.0/gems/rspec-core-3.3.1/exe/rspec --pattern spec/\*\*\{,/\*/\*\*\}/\*_spec.rb

Mcrain::Mysql
  .start
    ping
  start twice
    should not eq #<Mysql2::Client:0x00000001ee2158 @read_timeout=nil, @query_options={:as=>:hash, :async=>false, :cast_booleans=>false, :symbolize_keys=>false, :database_timezone=>:local, :application_timezone=>nil, :cache_rows=>true, :connect_flags=>2147525125, :cast=>true, :default_file=>nil, :default_group=>nil, :host=>"172.17.42.1", :port=>41453, :username=>"root"}, @current_query_options={:as=>:hash, :async=>false, :cast_booleans=>false, :symbolize_keys=>false, :database_timezone=>:local, :application_timezone=>nil, :cache_rows=>true, :connect_flags=>2147525125, :cast=>true, :default_file=>nil, :default_group=>nil, :host=>"172.17.42.1", :port=>41453, :username=>"root"}>
  skip_reset_after_teardown
    false
    true

Mcrain::Rabbitmq
  .start
[Faraday::ConnectionFailed] Connection refused - connect(2) for "172.17.42.1" port 45717
[Faraday::ConnectionFailed] Connection refused - connect(2) for "172.17.42.1" port 45717
[Faraday::ConnectionFailed] Connection refused - connect(2) for "172.17.42.1" port 45717
    overview
    should eq "http"
    should eq "rabbitmq"
  start twice
    should not eq #<RabbitMQ::HTTP::Client:0x00000003323c20 @endpoint="http://172.17.42.1:40330", @options={}, @connection=#<Faraday::Connection:0x00000003323608 @parallel_manager=nil, @headers={"Authorization"=>"Basic Z3Vlc3Q6Z3Vlc3Q=", "User-Agent"=>"Faraday v0.9.1"}, @params={}, @options=#<Faraday::RequestOptions (empty)>, @ssl=#<Faraday::SSLOptions (empty)>, @default_parallel_manager=nil, @builder=#<Faraday::RackBuilder:0x00000003323388 @handlers=[FaradayMiddleware::FollowRedirects, Faraday::Response::RaiseError, FaradayMiddleware::ParseJson, Faraday::Adapter::NetHttp], @app=#<FaradayMiddleware::FollowRedirects:0x00000003322078 @app=#<Faraday::Response::RaiseError:0x000000033220c8 @app=#<FaradayMiddleware::ParseJson:0x00000003322140 @app=#<Faraday::Adapter::NetHttp:0x00000003322190 @app=#<Proc:0x00000003322258@/home/yamada/go/src/github.com/groovenauts/mcrain/vendor/bundle/ruby/2.1.0/gems/faraday-0.9.1/lib/faraday/rack_builder.rb:152 (lambda)>>, @options={:content_type=>/\bjson$/}, @content_types=[/\bjson$/]>>, @options={:limit=>3}, @replay_request_codes=#<Set: {307}>>>, @url_prefix=#<URI::HTTP:0x000000033230e0 URL:http://172.17.42.1:40330/api>, @proxy=nil>>
  don't reset for first start
    should not eq "http://guest:guest@172.17.42.1:34095"

Mcrain::Redis
  .start
    ping
  start twice
    should not eq #<Redis client v3.2.1 for redis://172.17.42.1:57486/0>
  skip_reset_after_teardown
    false
    true

Mcrain::Riak
  .start
    should not eq #<Riak::Client [#<Node 172.17.42.1:55278>]>
  don't reset for first start
    should be truthy
  allow duplicated
    example at ./spec/mcrain/riak_spec.rb:40 (FAILED - 1)

Mcrain
  has a version number
  .[]
    should be a kind of Mcrain::Redis
    should eq #<Mcrain::Redis:0x000000033e4218 @port=50726, @container=#<Docker::Container:0x000000036e1628 @id="3c5294af792882d78cf2e16c8e8707a495540657257aa34f34c74b13ad90861b", @info={"Warnings"=>nil, "id"=>"3c5294af792882d78cf2e16c8e8707a495540657257aa34f34c74b13ad90861b"}, @connection=#<Docker::Connection:0x000000020564a8 @url="tcp://172.17.42.1:2375", @options={}>>, @host="172.17.42.1", @client=#<Redis client v3.2.1 for redis://172.17.42.1:50726/0>, @skip_reset_after_teardown=nil, @url="redis://172.17.42.1:50726">
WARNING: Using the `raise_error` matcher without providing a specific error or message risks false positives, since `raise_error` will match when Ruby raises a `NoMethodError`, `NameError` or `ArgumentError`, potentially allowing the expectation to pass without even executing the method you are intending to call. Instead consider providing a specific error class or message. This message can be supressed by setting: `RSpec::Expectations.configuration.warn_about_potential_false_positives = false`. Called from /home/yamada/go/src/github.com/groovenauts/mcrain/spec/mcrain_spec.rb:11:in `block (3 levels) in <top (required)>'.
    should raise Exception

Failures:

  1) Mcrain::Riak allow duplicated 
     Failure/Error: Mcrain[:riak].start do |s0|
     Docker::Error::NotFoundError:
       no such id: 7cba644639e56341cb7d8a385befa3fc590aaa980bd319d96e933ef3f8c8b08f
     # ./vendor/bundle/ruby/2.1.0/gems/docker-api-1.21.4/lib/docker/connection.rb:46:in `rescue in request'
     # ./vendor/bundle/ruby/2.1.0/gems/docker-api-1.21.4/lib/docker/connection.rb:38:in `request'
     # ./vendor/bundle/ruby/2.1.0/gems/docker-api-1.21.4/lib/docker/connection.rb:65:in `block (2 levels) in <class:Connection>'
     # ./vendor/bundle/ruby/2.1.0/gems/docker-api-1.21.4/lib/docker/container.rb:157:in `start!'
     # ./lib/mcrain/riak.rb:134:in `each'
     # ./lib/mcrain/riak.rb:134:in `setup'
     # ./lib/mcrain/base.rb:28:in `start'
     # ./spec/mcrain/riak_spec.rb:41:in `block (3 levels) in <top (required)>'

Finished in 2 minutes 23.7 seconds (files took 0.29362 seconds to load)
20 examples, 1 failure

Failed examples:

rspec ./spec/mcrain/riak_spec.rb:40 # Mcrain::Riak allow duplicated 

/usr/bin/ruby2.1 -I/home/yamada/go/src/github.com/groovenauts/mcrain/vendor/bundle/ruby/2.1.0/gems/rspec-core-3.3.1/lib:/home/yamada/go/src/github.com/groovenauts/mcrain/vendor/bundle/ruby/2.1.0/gems/rspec-support-3.3.0/lib /home/yamada/go/src/github.com/groovenauts/mcrain/vendor/bundle/ruby/2.1.0/gems/rspec-core-3.3.1/exe/rspec --pattern spec/\*\*\{,/\*/\*\*\}/\*_spec.rb failed
```

But only 1 example "Mcrain::Riak allow duplicated" succeeded.

```
yamada@MIRA% bundle exec rspec ./spec/mcrain/riak_spec.rb:39
Run options: include {:locations=>{"./spec/mcrain/riak_spec.rb"=>[39]}}

Mcrain::Riak
  allow duplicated
    should be truthy

Finished in 1 minute 32.09 seconds (files took 0.26551 seconds to load)
1 example, 0 failures

```

**Comment by akm** [comment]
@nagachika Thank you for your advice.
I added setup descriptions for docker to README.md. Please check them.

File: lib/mcrain/riak.rb (hunk 1 above)
Comment (akm) [inline lib/mcrain/riak.rb]: I fixed this mistake.

**Comment by akm** [comment]
@minimum2scp Thank you for your reporting.
The instance which was created by `Mcrain[:redis]` kept skip_reset_after_teardown, so I fixed it to clear skip_reset_after_teardown after run example.

**Pull request merged**